spec/u-namespace.md

```diff
@@ -0,0 +1,65 @@
+# MessageFormat 2.0 Unicode Namespace
+
+The `u:` namespace is reserved for use by the Unicode Consortium.
+
+## Options
+
+This section describes common **_<dfn>`u:` options</dfn>_** which each implementation SHOULD support
+for all _functions_ and _markup_.
+
+### `u:id`
+
+A string value that is included as an `id` or other suitable value
+in the formatted parts for the _placeholder_,
+or any other structured formatted results.
+
+Ignored when formatting a message to a string.
+
+Accepts string values, or values which can be stringified without error.
+For other values, a _Bad Option_ error is emitted
+and the `u:id` option is ignored.
+
+### `u:locale`
+
+A comma-delimited list of BCP 47 language tags,
+or an implementation-defined list of such tags.
+
+Replaces the _locale_ defined in the _function context_ for this _expression_.
+The value is ignored when set on _markup_.
+
+During processing, the `u:locale` option
+is always removed from the resolved mapping of _options_.
+
+Values matching the following ABNF are always accepted:
+```abnf
+u-locale-option = langtag *([s] "," [s] langtag)
```

Comment:
`[s]` is the same as `o`. Use that.

I don't agree with using `langtag` over using `Language-Tag`. Grandfathered tags all have modern replacements (or are deprecated outright) and private use is the user's problem. There's no reason for us to inject additional checking gorp onto the language tag space.

Reply:
```suggestion
u-locale-option = langtag *(o "," o langtag)
```

Reply:
> There's no reason for us to inject additional checking gorp onto the language tag space.

I agree with this sentiment, but I think the current language is effectively required to allow for that. Note that we also include this further below:

> Implementations MAY support additional language tags,
> such as private-use or grandfathered tags,
> or tags using `_` instead of `-` as a separator.

With `langtag` as a base, we allow but do not require grandfathered or private-use tags to be supported, while also allowing but not requiring Unicode locale identifiers to be supported.

The overall intent here is to require the intersection of BCP47 and ULI to be supported, and allow extension from there, so that each implementation can use whatever is its native language tag format, and doesn't need to inject additional checking gorp onto its language tag space.

Reply:
My tendency has been to require "well-formed" language tags (including allowing grandfathered and private-use) but encourage "valid" tags. Does the following capture the intention:

Implementations:
- MUST support well-formed language tags
- SHOULD check for valid language tags
- SHOULD canonicalize language tags and SHOULD ensure that tags are ULIs
- MAY omit, reject, or skip individual invalid tags
- MAY support additional implementation-defined identifiers
- MAY support implementation-defined locale objects
- MUST omit language tags from the resolved value

Users:
- SHOULD use valid canonicalized ULIs

I notice in making this list that we don't say whether the `language` attribute of the formatted value should be modified to match the selected `u:locale`.

Reply:
> MUST support well-formed language tags

This is problematic at least in a JS context, as it would require that e.g. a grandfathered tag like `i-klingon` is supported, and it currently is not: https://tc39.es/ecma402/#sec-language-tags

It is not reasonable to require an implementation to extend or modify its language tag rules to accommodate MF2.

Reply:
You're right. Support is the wrong word, at least from the point of view of "locales that do something useful". My proposal should probably say "MUST accept well-formed language tags". (Note that canonicalization would turn `i-klingon` into `tlh`). The other SHOULD/MAY requirements allow well-formed but invalid tags to not be supported (consistent with EMCA-402). But a naive implementation could exist that accepted any seemingly-possible language tag (`xxx-Yyyy-ZZ-variant`)

Reply:
I would stongly suggest referencing instead https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id

It has a slightly narrower syntax than BCP47; see https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#bcp-47-conformance

(Note: once the spec goes final, these links will be modified to be of the form https://www.unicode.org/reports/tr35/xxx/tr35.html#yyy

Reply:
Referencing [`unicode_bcp47_locale_id`](https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id) would be fine as well; it looks like it's effectively the same as the current `langtag`?

Reply:
"MUST accept" means that any well-formed language tags can be passed in the language priority list. The range error in Intl happens later--it's allowed by "MAY omit, reject, or skip". The point is to be somewhat permissive about passing in things that might be language tags and then allowing implementations some leeway in disposing of them. In practice, no one is using `i-klingon` or `i-default` or `cel-gaulish`.

Reply:
Now I'm a bit confused. With what you propose, would it be acceptable for an implementation to emit a Bad Option error if `i-klingon` is used as a `u:locale` value?

Reply:
Yes, although possibly not fatally.

The point of these attributes is to set the locale for some formatting. Ideally, we should allow any value that can appear in e.g. HTML's `lang` attribute. But we can't enforce that such a locale "works" or has data installed on a given implementation or system. So if you pass `u:locale=cel-gaulish` (or `u:locale=$celGaulish`), you probably get Bad Option and a formatted message using whatever the contextual locale is. FWIW, an implementation might do the same for non-range-erroring `aaa-Yyyy-RR` (not valid) or even the valid tag `tlh-Cyrl-AQ`. All of the canonicalization stuff we point to is in the service of making tags around the margins marginally useful.

Reply:
What I propose is that 

1. There be a [ VC: ] constraint that the syntax for locale identifiers conforms to BCP47 so that we don't get complete garbage.
2. Implementations should disregard any locale identifier that they don't support (that is, not raise an error)
3. Implementations should follow the semantics of locale identifiers as in https://www.unicode.org/reports/tr35/tr35-general.html#unicode_bcp47_locale_id  
    - **NB: when moved to the CLDR spec, this should be a relative link tr35.html#unicode_bcp47_locale_id**

Reply:
I can live with that.

Reply:
@eemeli ?

Reply:
As I mentioned before, I'd be fine referencing `unicode_bcp47_locale_id` instead of `langtag`, but I'm not really sure how the further proposed changes would affect the text.

We already do not consider errors on options to be fatal, so I'd much rather not further restrict implementations from emitting errors for unsupported locales.

I'm going to apply an update, and then ask for another review pass.